### PR TITLE
Cleanup darwin usage of strtok and strncpy in DNS implementation

### LINKED ISSUE #11884: Use after free in DnssdImpl.cpp darwin code.
#### Problem
    char * tokens  = strdup(type);
    char * regtype = strtok(tokens, ".");
    free(tokens);

Current code has ^. strtok doesn't create a copy of tokens - it just creates a pointer. Free should happen at the end. But actually we should really be using a null checked safe string copy (MemoryAllocString / MemoryFree) and checking for null. Also strtok is not reentrant.

 #### Proposed Solution
option 1) Given that this platform impl already uses vector, why not use a c++ string?
option 2) strtok -> strtok_r, use MemoryAllocString, add check that tokens != null, move free to end.

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
- Use after free: strtok returns a pointer into a string, so trying to use the falue after a `free` is wrong
- unchecked (and unneeded) strdup usage (may return nullptr)
- replaced strncpy with CopyString (chip implementation guarantees null terminator)

#### Change overview
- use CopyString
- do an in-place change to truncate the type to the first . character.

#### Testing
Expect CI to validate the changes (I made the changes on a linux machine)

### CODE REVIEW

File: src/platform/Darwin/DnssdImpl.cpp [hunk 1]
```diff
@@ -426,7 +429,7 @@ static CHIP_ERROR GetAddrInfo(void * context, DnssdResolveCallback callback, uin
         err = TXTRecordGetItemAtIndex(txtLen, txtRecord, i, kDnssdKeyMaxSize, key, &valueLen, &valuePtr);
         VerifyOrReturnError(CheckForSuccess(sdCtx, __func__, err, true), CHIP_ERROR_INTERNAL);
 
-        strncpy(value, reinterpret_cast<const char *>(valuePtr), valueLen);
+        Platform::CopyString(value, reinterpret_cast<const char *>(valuePtr));
```
Comment (bzbarsky-apple) [inline src/platform/Darwin/DnssdImpl.cpp:432]: I don't think `valuePtr` is null-terminated here.  This should be more like:
```suggestion
        CopyString(value, ByteSpan(static_cast<const uint8_t *>(valuePtr), valueLen));
```
or so.  And probably need to #include Span.h in this file for that.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/platform/Darwin/DnssdImpl.cpp (hunk 1 above)
Comment (andy31415) [reply src/platform/Darwin/DnssdImpl.cpp]: Updated to use memcpy. I thought usage of strncpy and cast to const char * means this was a null terminated string. But if this is just memory, memcpy seems apropriate and no casts needed.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/platform/Darwin/DnssdImpl.cpp [hunk 2]
```diff
@@ -426,7 +429,15 @@ static CHIP_ERROR GetAddrInfo(void * context, DnssdResolveCallback callback, uin
...
-        strncpy(value, reinterpret_cast<const char *>(valuePtr), valueLen);
+        if (valueLen >= sizeof(value))
+        {
+            // Truncation, but nothing better we can do
+            valueLen = sizeof(value) - 1;
+        }
+        if (valueLen > 1)
+        {
+            memcpy(value, valuePtr, valueLen - 1);
+        }
```
Comment (bzbarsky-apple) [inline src/platform/Darwin/DnssdImpl.cpp:440]: I don't get this part.  Shouldn't this just be:
```suggestion
          memcpy(value, valuePtr, valueLen);
```
?  We already made sure that `valueLen` leaves 1 byte of space in `value`; why would we only copy `valueLen - 1` bytes?

**Review by bzbarsky-apple — DISMISSED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

**Comment by andy31415** [comment]
fast track: CI passes (finally) and was checked by platform owner.

**Pull request merged**